src/grpc_metrics.rs

```diff
@@ -0,0 +1,92 @@
+use std::{sync::{Arc, RwLock}, time::Instant};
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+use rocket::http::hyper;
+use rocket::http::hyper::Body;
+use rocket::log;
+use tonic::body::BoxBody;
+use tower::{Layer, Service};
+
+use crate::metrics::PrometheusMetrics;
+
+#[derive(Clone)]
+pub struct GrpcMetricLayer {
+    prometheus: Arc<RwLock<PrometheusMetrics>>,
+}
+
+impl GrpcMetricLayer {
+    pub(crate) fn new(prom: Arc<RwLock<PrometheusMetrics>>) -> Self {
+        Self {
+            prometheus: prom
+        }
+    }
+}
+
+impl<S> Layer<S> for GrpcMetricLayer {
+    type Service = GrpcMetric<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        GrpcMetric { inner: service, prometheus: Arc::clone(&self.prometheus) }
+    }
+}
+
+#[derive(Clone)]
+pub struct GrpcMetric<S> {
+    inner: S,
+    prometheus: Arc<RwLock<PrometheusMetrics>>,
+}
+
+type BoxFuture<'a, T> = Pin<Box<dyn std::future::Future<Output=T> + Send + 'a>>;
+
+impl<S> Service<hyper::Request<Body>> for GrpcMetric<S>
+    where
+        S: Service<hyper::Request<Body>, Response=hyper::Response<BoxBody>> + Clone + Send + 'static,
+        S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: hyper::Request<Body>) -> Self::Future {
+        let clone = self.inner.clone();
+        let mut inner = std::mem::replace(&mut self.inner, clone);
+        let prom = self.prometheus.read().unwrap().clone();
```

Comment:
You can use `tokio::sync::RwLock` instead of `std::sync::RwLock` to get around this unwrap.

Reply:
Actually, I tried to use this, but it caused following error:
`returning this value requires that `'1` must outlive `'static``

Do you have any solution for this?

Reply:
I'm not sure it's better but you can do like this:

```rust
let prom = self.prometheus.clone();
Box::pin(async move {
  let prom = prom.read().await;
}
``` 

Reply:
yeah, it works